torch/utils/checkpoint.py

```diff
@@ -1514,7 +1518,12 @@ def recompute_fn(*inputs):
             device_autocast_ctx = torch.amp.autocast(
                 device_type=device_type, **device_autocast_kwargs
             ) if torch.amp.is_autocast_available(device_type) else contextlib.nullcontext()
-            with device_autocast_ctx, torch.amp.autocast("cpu", **cpu_autocast_kwargs), recompute_context:  # type: ignore[attr-defined]
+            with (
+                device_autocast_ctx,   # type: ignore[attr-defined]
+                torch.amp.autocast("cpu", **cpu_autocast_kwargs),
+                _apply_torch_function_mode_stack(torch_function_mode_stack),
```

Comment:
noob question - during the recompute phase, are we effectively rerunning the user's forward function (and all of there `torch.*` ops)? Or is the AC code here doing something different, like capturing all of the ATen ops that is witnessed during the forward, and replaying those. If it's the latter, I'm not sure if it will work as cleanly with a TorchFunctionMode

Reply:
in eager, its the former - we're just rerunning the user's forward function

for compile, I'm not sure what the status of TorchFunctionMode is with compile generally, but assuming that the TorchFunctionMode is traced through by dynamo, I guess the subgraph that the HOP applies the is_recompute annotations to should include the TorchFunctionMode logic baked in

Reply:
yep I think this is right

Reply:
Yes, this is correct @soulitzer